### ISSUE TITLE
Error in pcscf

### ISSUE THREAD

**Issue by ghost** [herlesupreeth/Open5gs_Config#5]
Dear @herlesupreeth 
Hi again

As u guided i successfully connected UE (samsung A20) to the Core and i have internet connection
But i do not see any VOLTE sign, while my phone support it!

on the other hand i Have error on PCSCF!

which says 

`98 (7000) ERROR: <script>: preloading NAT-PING. Rows:0

while the other IMS componets are not showing Errors after run the FOHSS   

Just written 

accept connection
Add serviced peer
Dropping Service peer
`

Would u please help me to solve this problem?

Thanks 
Best regards

**Comment by herlesupreeth** [comment]
Please attach a trace in order for me to help you solve the issue

> `98 (7000) ERROR: <script>: preloading NAT-PING. Rows:0

This is not an error, its just printing logs using ERROR level

**Comment by Mopoma95** [comment]
These Are my OpenStack screenshots:
[
![Screenshot (1)](https://user-images.githubusercontent.com/38971099/100875335-e0109a80-3473-11eb-8f32-224ab0f2c1bb.png)
![Screenshot (2)](https://user-images.githubusercontent.com/38971099/100875340-e1da5e00-3473-11eb-9cb0-190ec969277e.png)
![Screenshot (3)](https://user-images.githubusercontent.com/38971099/100875342-e272f480-3473-11eb-8148-90efe59e8276.png)
![Screenshot (4)](https://user-images.githubusercontent.com/38971099/100875345-e3a42180-3473-11eb-85be-ef184a90d6d3.png)
![Screenshot (5)](https://user-images.githubusercontent.com/38971099/100875348-e43cb800-3473-11eb-9a3c-a448b3ccfce8.png)
![Screenshot (6)](https://user-images.githubusercontent.com/38971099/100875349-e4d54e80-3473-11eb-8a5b-ce14575283ff.png)
![Screenshot (7)](https://user-images.githubusercontent.com/38971099/100875351-e56de500-3473-11eb-9acb-2253c821b219.png)
![Screenshot (8)](https://user-images.githubusercontent.com/38971099/100875356-e6067b80-3473-11eb-9d19-07efbf782878.png)
![Screenshot (9)](https://user-images.githubusercontent.com/38971099/100875358-e69f1200-3473-11eb-9782-4780ebb96efc.png)
![Screenshot (10)](https://user-images.githubusercontent.com/38971099/100875360-e868d580-3473-11eb-8ea7-aa376b3d8614.png)
![Screenshot (11)](https://user-images.githubusercontent.com/38971099/100875361-e9016c00-3473-11eb-9edd-f7fce415fb4b.png)
![Screenshot (12)](https://user-images.githubusercontent.com/38971099/100875366-ea329900-3473-11eb-9658-159d047c62ce.png)











](url)




And These are my SRSLTE screen shots:



![Screenshot from 2020-12-02 07-35-19](https://user-images.githubusercontent.com/38971099/100875587-354cac00-3474-11eb-81e3-a7e9e0190d9f.png)
![Screenshot from 2020-12-02 07-35-29](https://user-images.githubusercontent.com/38971099/100875591-367dd900-3474-11eb-9406-63ab91ae77d7.png)
![Screenshot from 2020-12-02 07-35-33](https://user-images.githubusercontent.com/38971099/100875593-37166f80-3474-11eb-893f-9cbefa6d2b2b.png)

**Comment by Mopoma95** [comment]
> Please attach a trace in order for me to help you solve the issue
> 
> > `98 (7000) ERROR: <script>: preloading NAT-PING. Rows:0
> 
> This is not an error, its just printing logs using ERROR level



Anyway I do not have VOLTE on my device!

**Comment by herlesupreeth** [comment]
As I see in your logs, UE is not even trying to attach for IMS APN. You need to enable VoLTE on the phone first

**Comment by Mopoma95** [comment]
I did it,
and Also set IMS APN but in when I select IMS APN it doesnt have internet connectivity! neither VOLTE

**Comment by herlesupreeth** [comment]
Can you describe how did you activate VoLTE on that phone?

**Comment by Mopoma95** [comment]
I will send u my configs:

![Screenshot from 2020-12-02 08-10-38](https://user-images.githubusercontent.com/38971099/100877037-3da5e680-3476-11eb-89dc-5e934172bf09.png)
![Screenshot from 2020-12-02 08-10-47](https://user-images.githubusercontent.com/38971099/100877040-3f6faa00-3476-11eb-8bc2-3ca45b29a28e.png)
![Screenshot from 2020-12-02 08-10-55](https://user-images.githubusercontent.com/38971099/100877043-40a0d700-3476-11eb-833b-7fcad9cef757.png)
![Screenshot from 2020-12-02 08-11-07](https://user-images.githubusercontent.com/38971099/100877046-41396d80-3476-11eb-80f2-768f06294494.png)
![Screenshot from 2020-12-02 08-11-20](https://user-images.githubusercontent.com/38971099/100877047-41d20400-3476-11eb-81df-284d05fa886c.png)
![Screenshot from 2020-12-02 08-12-10](https://user-images.githubusercontent.com/38971099/100877052-43033100-3476-11eb-8fe8-6bea7ee5abac.png)
![Screenshot from 2020-12-02 08-12-33](https://user-images.githubusercontent.com/38971099/100877060-439bc780-3476-11eb-8650-55878679830a.png)
![Screenshot from 2020-12-02 08-12-40](https://user-images.githubusercontent.com/38971099/100877064-44ccf480-3476-11eb-9c09-c96e4aeb25cf.png)
![Screenshot from 2020-12-02 08-12-46](https://user-images.githubusercontent.com/38971099/100877070-45658b00-3476-11eb-8e2d-53546accd904.png)
![Screenshot from 2020-12-02 08-12-51](https://user-images.githubusercontent.com/38971099/100877071-45fe2180-3476-11eb-963d-f824a16fc0dc.png)
![Screenshot from 2020-12-02 08-12-53](https://user-images.githubusercontent.com/38971099/100877074-4696b800-3476-11eb-9f55-ae4402047a41.png)

**Comment by Mopoma95** [comment]
> Can you describe how did you activate VoLTE on that phone?

Sure,
Go to settings
Connections>> Mobile Networks >> VOLTE CALLS SIM1 >> ENABLED

**Comment by herlesupreeth** [comment]
Those steps are not enough. Even though you toggle VoLTE on the phone settings it wont be activated at firmware level. Refer this link on how to activate VoLTE using USIM - https://github.com/herlesupreeth/CoIMS_Wiki

**Comment by Mopoma95** [comment]
Thanks,
I will Try to enable it,


Are those configs and other Parameters right?

**Comment by herlesupreeth** [comment]
I went through briefly and settings looked right. Rather than these screenshots a wireshark trace would be much better to debug

**Comment by Mopoma95** [comment]
Thanks,

I do not have to access for buying Gemalto or other programmer!
I bout this one : [](https://www.amazon.com/Reader-Writer-Programable-Version-Personalize/dp/B08FG5BL3M)

I just have this :


![Screenshot 2020-12-02 132640](https://user-images.githubusercontent.com/38971099/100878678-bed59b80-34bf-11eb-842a-7d016f9d1dbf.png)
![Screenshot 2020-12-02 132502](https://user-images.githubusercontent.com/38971099/100878683-c006c880-34bf-11eb-83ac-4343d0deca13.png)
![Screenshot 2020-12-02 132621](https://user-images.githubusercontent.com/38971099/100878686-c09f5f00-34bf-11eb-944a-2c3a3859f0e4.png)




Would u please check if this Programmer can handle COIMS tasks? and please check my programmed sim with those setups in the pic


Thanks 
Best regards

**Comment by Mopoma95** [comment]
The Java tab is disabled!
I may contact to the producer but could u please check other specs?

**Comment by Mopoma95** [comment]
Dear @herlesupreeth 

I Just Noted that i should set apn type for ims to "ims"


I did set that and the Turend off the instance in openstack and srslte , and the run them all again

These are new LOGS
Would u please check that?

![Screenshot (13)](https://user-images.githubusercontent.com/38971099/100883342-3be02100-347e-11eb-996a-0154a68feddf.png)
![Screenshot (14)](https://user-images.githubusercontent.com/38971099/100883348-3da9e480-347e-11eb-8ee8-1ebce8538953.png)
![Screenshot (15)](https://user-images.githubusercontent.com/38971099/100883352-3e427b00-347e-11eb-8546-3fbb2e2c46b1.png)
![Screenshot (16)](https://user-images.githubusercontent.com/38971099/100883356-3edb1180-347e-11eb-8b58-f9e2d1191238.png)
![Screenshot (17)](https://user-images.githubusercontent.com/38971099/100883360-3f73a800-347e-11eb-8477-517ecd5410ba.png)
![Screenshot (18)](https://user-images.githubusercontent.com/38971099/100883363-400c3e80-347e-11eb-8a41-cdb8a8f3469e.png)
![Screenshot (19)](https://user-images.githubusercontent.com/38971099/100883367-40a4d500-347e-11eb-93a8-80ea7b42cead.png)
![Screenshot (20)](https://user-images.githubusercontent.com/38971099/100883369-413d6b80-347e-11eb-98ad-93629a4b4230.png)
![Screenshot (21)](https://user-images.githubusercontent.com/38971099/100883371-41d60200-347e-11eb-897f-e4edf51bf345.png)

**Comment by herlesupreeth** [comment]
Looks like your UE is trying to attach to IMS now. That's good news. Please dont spam here with screenshots, rather send wireshark traces

**Comment by Mopoma95** [comment]
Sorry For spamming out

This is Wireshark
[Volte_not_Shown_ON_UE.zip](https://github.com/herlesupreeth/Open5gs_Config/files/5629940/Volte_not_Shown_ON_UE.zip)


I could trace wireshark from where SRSLTE was, I mean from native ubuntu not openStack

**Comment by herlesupreeth** [comment]
I believe you have not added the IMS subscription details in FHoSS, see the screenshot below from the trace you shared

![image](https://user-images.githubusercontent.com/4115839/100888427-64840d00-34b6-11eb-9814-3b59dd2b9a86.png)

**Comment by Mopoma95** [comment]
Sorry for uploading PICs,
I will delete most of them on successful result


![Screenshot from 2020-12-02 10-14-32](https://user-images.githubusercontent.com/38971099/100891649-7e5a2b80-3487-11eb-963c-2ba97928258c.png)
![Screenshot from 2020-12-02 10-14-46](https://user-images.githubusercontent.com/38971099/100891654-7f8b5880-3487-11eb-9e9f-c2923285ab59.png)
![Screenshot from 2020-12-02 10-15-30](https://user-images.githubusercontent.com/38971099/100891658-8023ef00-3487-11eb-8afd-7eaffb795fb2.png)


Thanks

**Comment by Mopoma95** [comment]
Sorry for uploading images
I will delete them on successful results


![Screenshot from 2020-12-02 10-14-32](https://user-images.githubusercontent.com/38971099/100891722-9336bf00-3487-11eb-8cd3-5741c5f59592.png)
![Screenshot from 2020-12-02 10-14-46](https://user-images.githubusercontent.com/38971099/100891727-9467ec00-3487-11eb-9e49-f472e17c225f.png)
![Screenshot from 2020-12-02 10-15-30](https://user-images.githubusercontent.com/38971099/100891729-95008280-3487-11eb-82cf-34b0cbdb1c5c.png)


Thank you so much

**Comment by Mopoma95** [comment]
172.24.16.3 is my UE ip address,
My SRSLTE device has 192.168.100.7 ip address
and the instance of openstack has 192.168.100.197 for floating ip and 10.4.128.21 for local ip
and the PC which has openstacked installed on it has 192.168.100.3 and 192.168.100, 192.168.100.4 with virbr0 192.168.122.1


should i set this 172.24.16.0/24 some where in HSS or FOHSS?

**Comment by Mopoma95** [comment]
> I believe you have not added the IMS subscription details in FHoSS, see the screenshot below from the trace you shared
> 
> ![image](https://user-images.githubusercontent.com/4115839/100888427-64840d00-34b6-11eb-9814-3b59dd2b9a86.png)

172.24.16.3 is my UE ip address,
My SRSLTE device has 192.168.100.7 ip address
and the instance of openstack has 192.168.100.197 for floating ip and 10.4.128.21 for local ip
and the PC which has openstacked installed on it has 192.168.100.3 and 192.168.100, 192.168.100.4 with virbr0 192.168.122.1

should i set this 172.24.16.0/24 some where in HSS or FOHSS?

**Comment by Mopoma95** [comment]
instance has 4 interfaces:

ens3 10.4.128.21
lo 127.0.0.1
ogstun 172.24.15.1
ogstun2 172.24.16.1

**Comment by herlesupreeth** [comment]
I believe this screenshot belongs to your ICSCF, its looks FHoSS is not connected to ICSCF

https://user-images.githubusercontent.com/38971099/100883352-3e427b00-347e-11eb-8546-3fbb2e2c46b1.png

**Comment by Mopoma95** [comment]
I run wireshark again

This is the new one 
[VOlte.zip](https://github.com/herlesupreeth/Open5gs_Config/files/5630377/VOlte.zip)
![Screenshot from 2020-12-02 10-36-59](https://user-images.githubusercontent.com/38971099/100894793-dc3c4280-348a-11eb-942c-23bb8ed1622e.png)




The Image u mentioned is exactly ICSCF!
could u please see PCSCF image in this https://user-images.githubusercontent.com/38971099/100883371-41d60200-347e-11eb-897f-e4edf51bf345.png

there is an ERROR in PCSCF!

**Comment by Mopoma95** [comment]
What problem might happened?
I will run all the components again and will upload the pics, because i do not have access to take a log from openstack components!
Sorry for that

I'll sent it in about 2 min

**Comment by herlesupreeth** [comment]
Send me FHoSS logs and also IMS configuration files along with DNS zone file

**Comment by Mopoma95** [comment]
[FOHSS_CONF_LOG_pic.zip](https://github.com/herlesupreeth/Open5gs_Config/files/5630671/FOHSS_CONF_LOG_pic.zip)


Sorry for late, i couldnt copy logs because used ssh so oblighed to use cat instead of nano and copyed cat of conf files


Thanks in advance

**Comment by herlesupreeth** [comment]
There is something definitely wrong with your FHoSS setup and even the connection between PCRF and PCSCF as I see Diameter connection errors in your logs and screenshots

Can you send the output of following commands?

```
ps aux | grep open5gs*

netstat -tupln | grep 38

ip a

ip r

iptables -t nat -L

iptables -L

cat /proc/sys/net/ipv4/ip_forward
```

**Comment by Mopoma95** [comment]
Hi, Good morning


```
root@vo-i:~# ps aux | grep open5gs*
grep: open5gs-2.0.0: Is a directory
root@vo-i:~# ps aux | grep open5gs
root      6291  0.1  0.0 2520432 13528 pts/4   Sl+  05:37   0:00 ./open5gs/install/bin/open5gs-pcrfd
root      6335  0.0  0.0 923628 10512 pts/8    Sl+  05:37   0:00 ./open5gs/install/bin/open5gs-sgwd
root      6338  0.1  0.1 2596536 26784 pts/6   Sl+  05:38   0:00 ./open5gs/install/bin/open5gs-pgwd
root      6383  0.1  0.0 2464300 12540 pts/5   Sl+  05:38   0:00 ./open5gs/install/bin/open5gs-hssd
root      6422  0.2  0.0 2502592 11996 pts/0   Sl+  05:38   0:00 ./open5gs/install/bin/open5gs-mmed
root      6470  0.0  0.0  14864  1092 pts/9    S+   05:39   0:00 grep --color=auto open5gs
```

```
root@vo-i:~# netstat -tupln | grep 38
tcp        0      0 127.0.0.4:5868          0.0.0.0:*               LISTEN      6383/./open5gs/inst
tcp        0      0 127.0.0.3:5868          0.0.0.0:*               LISTEN      6338/./open5gs/inst
tcp        0      0 127.0.0.2:3868          0.0.0.0:*               LISTEN      6422/./open5gs/inst
tcp        0      0 127.0.0.4:3868          0.0.0.0:*               LISTEN      6383/./open5gs/inst
tcp        0      0 127.0.0.3:3868          0.0.0.0:*               LISTEN      6338/./open5gs/inst
tcp        0      0 127.0.0.5:3868          0.0.0.0:*               LISTEN      6291/./open5gs/inst
tcp        0      0 10.4.128.21:3869        0.0.0.0:*               LISTEN      6236/kamailio
tcp        0      0 10.4.128.21:3870        0.0.0.0:*               LISTEN      6185/kamailio
tcp        0      0 10.4.128.21:3871        0.0.0.0:*               LISTEN      6098/kamailio
tcp6       0      0 10.4.128.21:3868        :::*                    LISTEN      6245/java
udp        0      0 127.0.0.3:2123          0.0.0.0:*                           6338/./open5gs/inst
udp        0      0 127.0.0.3:2152          0.0.0.0:*                           6338/./open5gs/inst
udp6       0      0 ::1:2123                :::*                                6338/./open5gs/inst
udp6       0      0 ::1:2152                :::*                                6338/./open5gs/inst
```





```
root@vo-i:~# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: ens3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1450 qdisc fq_codel state UP group default qlen 1000
    link/ether fa:16:3e:f0:a5:e4 brd ff:ff:ff:ff:ff:ff
    inet 10.4.128.21/24 brd 10.4.128.255 scope global dynamic ens3
       valid_lft 68658sec preferred_lft 68658sec
    inet6 fe80::f816:3eff:fef0:a5e4/64 scope link
       valid_lft forever preferred_lft forever
3: ogstun: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1400 qdisc fq_codel state UP group default qlen 500
    link/none
    inet 172.24.15.1/24 scope global ogstun
       valid_lft forever preferred_lft forever
    inet6 fe80::e18c:621e:8b34:cc72/64 scope link stable-privacy
       valid_lft forever preferred_lft forever
4: ogstun2: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1400 qdisc fq_codel state UP group default qlen 500
    link/none
    inet 172.24.16.1/24 scope global ogstun2
       valid_lft forever preferred_lft forever
    inet6 fe80::cbf6:937e:b48c:842c/64 scope link stable-privacy
       valid_lft forever preferred_lft forever
```




```
root@vo-i:~# ip r
default via 10.4.128.1 dev ens3 proto dhcp src 10.4.128.21 metric 100
10.4.128.0/24 dev ens3 proto kernel scope link src 10.4.128.21
169.254.169.254 via 10.4.128.1 dev ens3 proto dhcp src 10.4.128.21 metric 100
172.24.15.0/24 dev ogstun proto kernel scope link src 172.24.15.1
172.24.16.0/24 dev ogstun2 proto kernel scope link src 172.24.16.1
```







```
root@vo-i:~# iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination

Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination
MASQUERADE  all  --  172.24.15.0/24       anywhere
```






```
root@vo-i:~# iptables -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination
ACCEPT     all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere
rtpengine  udp  --  anywhere             anywhere

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

Chain rtpengine (1 references)
target     prot opt source               destination
RTPENGINE  udp  --  anywhere             anywhere             RTPENGINE id:0
```


```
root@vo-i:~# cat /proc/sys/net/ipv4/ip_forward
1
```

**Comment by herlesupreeth** [comment]
On the machine with EPC + IMS, run only the following components and send me a wireshark trace for each case

1. Run only PCRF + PCSCF
2. Run only SCSCF + ICSCF + FHoSS

**Comment by Mopoma95** [comment]
How can i trace wireshark in openstack?

I could trace it with tshark but do not know how to send it to my PC which did ssh to the instance
Do u have any idea? or how to upload it somewhere through the net?

**Comment by herlesupreeth** [comment]
Install wireshark on OpenStack Vm and then you can run X11 forwarding as follows to OpenStack VM

ssh -X root@192.168.100.197

then run wireshark there in SSH session

**Comment by Mopoma95** [comment]
https://s17.picofile.com/file/8416172842/VOLTE.rar.html
please select the red box to download

**Comment by herlesupreeth** [comment]
please compress the wireshark trace as .zip and upload here in the issue

**Comment by Mopoma95** [comment]
I did compresses it but it got 25MB and git hub accept 10 MB of data!

**Comment by Mopoma95** [comment]
[VOLTE.zip](https://github.com/herlesupreeth/Open5gs_Config/files/5634405/VOLTE.zip)

Its just wireshark's u asked for and the file i put the link was the complete trace of All components

**Comment by herlesupreeth** [comment]
thanks for the trace, it looks like the link between  SCSCF + ICSCF + FHoSS is good, however the link between PCRF + PCSCF is not good. Please send the .xml files for PCSCF also which version of Ubuntu are you running? and which kernel version?

**Comment by Mopoma95** [comment]
Anytime,



```

root@vo-i:/etc/kamailio_pcscf# cat pcscf.xml
<?xml version="1.0" encoding="UTF-8"?>
<DiameterPeer
        FQDN="pcscf.ims.mnc001.mcc001.3gppnetwork.org"
        Realm="ims.mnc001.mcc001.3gppnetwork.org"
        Vendor_Id="10415"
        Product_Name="CDiameterPeer"
        AcceptUnknownPeers="1"
        DropUnknownOnDisconnect="1"
        Tc="30"
        Workers="4"
        QueueLength="8"
        TransactionTimeout="5"
        SessionsHashSize="128"
        DefaultAuthSessionTimeout="3600"
        MaxAuthSessionTimeout="3600"
>
        <Peer FQDN="pcrf.epc.mnc001.mcc001.3gppnetwork.org" Realm="epc.mnc001.mcc001.3gppnetwork.org" port="3868"/>

        <Acceptor port="3871" bind="10.4.128.21"/>

        <Auth id="16777236" vendor="10415"/> <!-- 3GPP Rx -->
        <Auth id="16777236" vendor="0"/> <!-- 3GPP Rx -->

        <DefaultRoute FQDN="pcrf.epc.mnc001.mcc001.3gppnetwork.org" metric="10"/>

</DiameterPeer>
```




```
root@vo-i:/etc/kamailio_pcscf# cat pcscf.cfg
# IP-Adress for incoming SIP-Traffic, in the following format:

# SIP / UDP
listen=udp:10.4.128.21:5060
# Uncomment the below line only when UE is behind double NAT (e.g. VoIP calling over WiFi/ CN behind a NAT)
#listen=udp:10.4.128.21:5060 advertise 172.24.15.30:5060
# SIP / TCP
listen=tcp:10.4.128.21:5060
# Uncomment the below line only when UE is behind double NAT (e.g. VoIP calling over WiFi/ CN behind a NAT)
#listen=tcp:10.4.128.21:5060 advertise 172.24.15.30:5060
# SIP / TCP/TLS
#listen=tls:11.22.33.44:5061

# IPSEC / UDP
#!define IPSEC_LISTEN_ADDR "10.4.128.21"
#!define IPSEC_CLIENT_PORT 5100
#!define IPSEC_SERVER_PORT 6100
#!define IPSEC_MAX_CONN 10

# IP used in Rx_AAR_Register - IP of this P-CSCF, to be used in the flow for the AF-signaling
#!define RX_AF_SIGNALING_IP "10.4.128.21"
# Uncomment the below line only when UE is behind double NAT (e.g. VoIP calling over WiFi/ CN behind a NAT)
##!define RX_AF_SIGNALING_IP "172.24.15.30"

alias=pcscf.ims.mnc001.mcc001.3gppnetwork.org

#!define MY_WS_PORT 80
#!define MY_WSS_PORT 443

#!define PCSCF_URL "sip:pcscf.ims.mnc001.mcc001.3gppnetwork.org:5060"

#!define TCP_PROCESSES 8

#!subst "/NETWORKNAME/ims.mnc001.mcc001.3gppnetwork.org/"
#!subst "/HOSTNAME/pcscf.ims.mnc001.mcc001.3gppnetwork.org/"
#!subst "/PCRF_REALM/epc.mnc001.mcc001.3gppnetwork.org/"

# SIP-Address of capturing node, if not set, capturing is disabled.
##!define CAPTURE_NODE "sip:127.0.0.1:9060"

# Allowed IPs for XML-RPC-Queries
##!define XMLRPC_WHITELIST_1 "127.0.0.1"
##!define XMLRPC_WHITELIST_2 "127.0.0.1"
##!define XMLRPC_WHITELIST_3 "127.0.0.1"

# Databases:
#!define DB_URL "mysql://pcscf:heslo@127.0.0.1/pcscf"
##!define DB_URL2 "con2=>mysql://pcscf:heslo@127.0.0.1/pcscf"

#!define SQLOPS_DBURL "pcscf=>mysql://pcscf:heslo@127.0.0.1/pcscf"

#! Optional: Server-URL for Websocket-Requests
##!define WEBSOCKET_WEBSERVER "phone.ng-voice.com"

##!define TRF_FUNCTION "trf.ims.mnc001.mcc001.3gppnetwork.org"

#
# Several features can be enabled using '#!define WITH_FEATURE' directives:
#
# *** To run in debug mode:
#     - define WITH_DEBUG
#
# *** To enable nat traversal execute:
#     - define WITH_NAT
#     - define the connection to the RTP-Proxy: RTPPROXY_ADDRESS
#
# *** To force alls calls through the RTP-Proxy
#     - this will automagically enable NAT-Traversal
#     - define FORCE_RTPRELAY
#
# *** To enable IPv4/IPv6 Translation (RTPProxy)
#     - this will automagically enable NAT-Traversal
#     - define WITH_RTPIPV4
#
# *** To enable TCP support execute:
#     - define WITH_TCP
#
# *** To enable TLS support execute:
#     - adjust CFGDIR/tls.cfg as needed
#     - define WITH_TLS
#     - this will automagically enable TCP
#
# *** To enable XMLRPC support execute:
#     - define WITH_XMLRPC
#     - this will automagically enable TCP
#
# *** To enable anti-flood detection execute:
#     - adjust pike and htable=>ipban settings as needed (default is
#       block if more than 16 requests in 2 seconds and ban for 300 seconds)
#     - define WITH_ANTIFLOOD
#
# *** To enable the Rx-Interface:
#     - Configure Rx-Diameter-Interface in pcscf.xml
#     - define WITH_RX
#
# *** To enable a Homer SIP-Capter-Node:
#     - define CAPTURE_NODE with a proper address
#
# *** To enable support for the SEMS-SBC
#     - define WITH_SBC
#     - configure dispatcher-list with a set of SBC's

#
# Enabled Features for this host:
##!define WITH_DEBUG
#!define WITH_NAT
#!define WITH_NATPING
#!define FORCE_RTPRELAY
##!define WITH_TLS
##!define WITH_XMLRPC
##!define WITH_IPBLOCK
##!define WITH_ANTIFLOOD
#!define WITH_RX
#!define WITH_RX_REG
#!define WITH_RX_CALL
#!define WITH_TCP
##!define WITH_RTPIPV4
##!define WITH_SBC
##!define WITH_SBC_CALL
##!define WITH_REGINFO
##!define WITH_RTPPING
##!define WITH_WEBSOCKET
#!define WITH_IPSEC
#!define WITH_IMS_HDR_CACHE
#!define WITH_PING_UDP
#!define WITH_PING_TCP
```










```
root@vo-i:/etc/kamailio_pcscf# cat kamailio_pcscf.cfg
#
# TelcoSuite (V3) Proxy-CSCF
#
# Kamailio (OpenSER) SIP Server
#     - web: http://www.kamailio.org
#     - git: http://sip-router.org
#
# Refer to the Core CookBook at http://www.kamailio.org/dokuwiki/doku.php
# for an explanation of possible statements, functions and parameters.
#

import_file "pcscf.cfg"

####### Defined Values #########
# *** Value defines - IDs used later in config
#!define DISPATCHER_LIST_SBC 1

#!define DISPATCHER_DST_AVP "i:1"
#!define DISPATCHER_GRP_AVP "i:2"
#!define DISPATCHER_CNT_AVP "i:3"
#!define DISPATCHER_SOCK_AVP "i:4"

#!define RR_CUSTOM_USER_AVP "i:5"
#!define DLG_TIMEOUT_AVP "i:6"

#!define FLT_MOBILE_ORIG 1
#!define FLT_DIALOG 2
#!define FLT_NAT 3
#!define FLT_RTP 4
#!define FLT_CAPTURE 5

####### Global Parameters #########

#!ifdef WITH_DEBUG
debug=4
log_stderror=yes
#!else
debug=2
log_stderror=no
#!endif

memdbg=5
memlog=5

max_while_loops=5000

#!ifdef WITH_XMLRPC
listen=tcp:127.0.0.1:5060
#!endif

# Locks all ser pages into memory making it unswappable (in general one
# doesn't want his sip proxy swapped out )
mlock_pages=yes
# Tries to pre-fault all the shared memory, before starting. When "on", start
# time will increase, but combined with mlock_pages will guarantee ser will get
# all its memory from the beginning (no more kswapd slow downs)
shm_force_alloc=yes

# Do SRV-Loadbalancing:
dns_srv_lb=yes
# Always prefer IPv6:
dns_try_ipv6=yes
# DNS-Based failover
use_dns_failover=yes
# Query NAPTR-Records as well:
dns_try_naptr=no

user_agent_header="User-Agent: TelcoSuite Proxy-CSCF"
server_header="Server: TelcoSuite Proxy-CSCF"

log_facility=LOG_LOCAL0

fork=yes
children=4

#!ifndef TCP_PROCESSES
# Number of TCP Processes
#!define TCP_PROCESSES 16
#!endif

#!ifdef WITH_TLS
# Check, if TCP is enabled:
#!ifndef WITH_TCP
#!define WITH_TCP
#!endif
enable_tls=yes
#!endif

#!ifdef WITH_XMLRPC
#!ifndef WITH_TCP
#!define WITH_TCP
#!endif
#!ifndef TCP_PROCESSES
# Number of TCP Processes
#!define TCP_PROCESSES 3
#!endif
#!endif

#!ifdef WITH_TCP
# life time of TCP connection when there is no traffic
# - a bit higher than registration expires to cope with UA behind NAT
tcp_connection_lifetime=36000
# If a message received over a tcp connection has "alias" in its via a new tcp
# alias port will be created for the connection the message came from (the
# alias port will be set to the via one).
#
# Note: For NAT traversal of TCP clients it is better to not use
# tcp_accept_aliases but just use nathelper module and
# fix_nated_[contact|register] functions.
tcp_accept_aliases=no
# Enable SIP outbound TCP keep-alive using PING-PONG (CRLFCRLF - CRLF).
tcp_crlf_ping=yes

tcp_reuse_port=yes

tcp_accept_no_cl=yes
tcp_rd_buf_size=16384

#!ifdef TCP_PROCESSES
tcp_children=TCP_PROCESSES
#!endif
#!else
disable_tcp=yes
#!endif

/* uncomment the next line to disable the auto discovery of local aliases
   based on reverse DNS on IPs (default on) */
auto_aliases=no

#phone2tel=1

udp_mtu = 1300
udp_mtu_try_proto = TCP

/* uncomment and configure the following line if you want Kamailio to
   bind on a specific interface/port/proto (default bind on all available) */

system.shutdownmode = 0 desc "System shutdown mode"
system.service = "Proxy-CSCF" desc "Function of this server"

####### Modules Section ########

# set paths to location of modules
mpath="/usr/lib64/kamailio/modules_k/:/usr/lib64/kamailio/modules/:/usr/lib/kamailio/modules_k/:/usr/lib/kamailio/modules/:/usr/lib/x86_64-linux-gnu/kamailio/modules/:/usr/local/lib64/kamailio/modules"

# Fifo Module
# Kamailio Extensions (e.g. MI:uptime, MI:version, cfg:isflagset etc.)
loadmodule "kex"
# Transaction Module
loadmodule "tm"
loadmodule "tmx"
loadmodule "sl"
loadmodule "rr"
loadmodule "pv"
loadmodule "maxfwd"
loadmodule "textops"
loadmodule "textopsx"
# SIP-Utilities: options_reply
loadmodule "siputils"
loadmodule "sanity"
loadmodule "ctl"
loadmodule "cfg_rpc"
loadmodule "xlog"
loadmodule "auth"
loadmodule "dispatcher"
loadmodule "sctp"
loadmodule "path"
loadmodule "statistics"

loadmodule "ims_dialog"
loadmodule "ims_usrloc_pcscf"
#!ifdef WITH_IPSEC
loadmodule "ims_ipsec_pcscf"
#!endif
loadmodule "ims_registrar_pcscf"

#!ifdef WITH_XMLRPC
loadmodule "xmlrpc"
#!endif

#!ifdef WITH_REGINFO
loadmodule "pua"
#!endif

#!ifdef DB_URL
loadmodule "db_mysql"
#!ifdef DB_URL2
loadmodule "db_cluster"
#!endif
#!endif

#!ifdef WITH_DEBUG
loadmodule "debugger"
#!endif

loadmodule "usrloc"
loadmodule "registrar"

loadmodule "nathelper"

#!ifdef WITH_ANTIFLOOD
loadmodule "pike"
#!endif

#!ifdef WITH_TLS
loadmodule "tls"
#!endif

#!ifdef WITH_RTPPING
loadmodule "rtpping"
#!endif


loadmodule "sdpops"
loadmodule "rtpengine"

#!ifdef WITH_WEBSOCKET
loadmodule "xhttp.so"
loadmodule "websocket.so"
#!endif

#!ifdef WITH_RX
loadmodule "cdp"
loadmodule "cdp_avp"
loadmodule "ims_qos"
#!endif

#!ifdef CAPTURE_NODE
loadmodule "siptrace"
#!endif

#!ifdef WITH_NATPING
loadmodule "rtimer"
loadmodule "uac"
loadmodule "sqlops"
#!endif

# HTable as a cache:
loadmodule "htable"

#!ifdef WITH_DEBUG
#loadmodule "debugger.so"
modparam("debugger", "mod_hash_size", 5)
modparam("debugger", "mod_level_mode", 1)
modparam("debugger", "mod_level", "rtpengine=3")
modparam("debugger", "mod_level", "ims_qos=3")
modparam("debugger", "mod_level", "ims_ipsec_pcscf=3")
modparam("debugger", "mod_level", "textops=3")
modparam("debugger", "mod_level", "tm=3")
modparam("debugger", "mod_level", "ims_registrar_pcscf=3")
modparam("debugger", "mod_level", "ims_usrloc_pcscf=3")
modparam("debugger", "cfgtrace", 1)
#!endif

loadmodule "jsonrpcs.so"
# ----- jsonrpcs params -----
modparam("jsonrpcs", "pretty_format", 1)
/* set the path to RPC fifo control file */
modparam("jsonrpcs", "fifo_name", "/var/run/kamailio_pcscf/kamailio_rpc.fifo")
/* set the path to RPC unix socket control file */
modparam("jsonrpcs", "dgram_socket", "/var/run/kamailio_pcscf/kamailio_rpc.sock")

# ----------------- setting module-specific parameters ---------------
#!ifdef DB_URL2
# ----- db_cluster params -----
modparam("db_cluster", "connection", DB_URL)
modparam("db_cluster", "connection", DB_URL2)
modparam("db_cluster", "cluster", "cluster1=>con1=2s2s;con2=1s1s")
#!endif

#!ifdef WITH_ANTIFLOOD
# ----- pike params -----
modparam("pike", "sampling_time_unit", 2)
modparam("pike", "reqs_density_per_unit", 16)
modparam("pike", "remove_latency", 4)

# ----- htable params -----
# ip ban htable with autoexpire after 5 minutes
modparam("htable", "htable", "ipban=>size=8;autoexpire=300")
modparam("htable", "htable", "failedauth=>size=8;autoexpire=120")
modparam("htable", "htable", "natpingfrom=>size=8;autoexpire=600000;")
#!endif

modparam("htable", "htable", "contact=>size=8;autoexpire=20")
modparam("htable", "htable", "a=>size=8;autoexpire=20")

#!ifdef WITH_IMS_HDR_CACHE
modparam("htable", "htable", "serviceroutes=>size=16;autoexpire=14400;")
modparam("htable", "htable", "associateduris=>size=16;autoexpire=14400;")
#!endif

#!ifdef WITH_NATPING
modparam("htable", "htable", "natping=>size=8;autoexpire=600000;")
modparam("htable", "htable", "natpingfail=>size=8;autoexpire=600000;")
modparam("htable", "htable", "natpingfrom=>size=8;autoexpire=600000;")
#!ifdef DB_URL2
modparam("sqlops","sqlcon","pcscf=>cluster://cluster1")
#!else
modparam("sqlops","sqlcon", SQLOPS_DBURL)
#!endif

modparam("uac","restore_mode","none")

# ----------------- Settings for RTimer ---------------
# time interval set to 60 seconds
modparam("rtimer", "timer", "name=NATPING;interval=60;mode=1;")
modparam("rtimer", "exec", "timer=NATPING;route=NATPING")
#!endif

# ----- tm params -----
# auto-discard branches from previous serial forking leg
#modparam("tm", "failure_reply_mode", 3)
# default retransmission timeout: 3 sec
modparam("tm", "fr_timer", 3000)
# default invite retransmission timeout after 1xx: 120sec
modparam("tm", "fr_inv_timer", 120000)
# Dont reply automatically with "100 Trying"
modparam("tm", "auto_inv_100", 0)

# ----- rr params -----
# add value to ;lr param to cope with most of the UAs
modparam("rr", "enable_full_lr", 1)
# do not append from tag to the RR (no need for this script)
modparam("rr", "append_fromtag", 1)
# add a Username to RR-Header
modparam("rr", "add_username", 1)
# Take User from a custom AVP
modparam("rr", "custom_user_avp", "$avp(RR_CUSTOM_USER_AVP)")

#!ifdef WITH_XMLRPC
# ----- xmlrpc params -----
modparam("xmlrpc", "route", "XMLRPC");
modparam("xmlrpc", "url_match", "^/RPC")
#!endif

#!ifdef WITH_TLS
# ----- tls params -----
modparam("tls", "config", "/etc/kamailio_pcscf/tls.cfg")
#!endif

# ----- rtpproxy params -----
modparam("rtpengine", "setid_default", 1)
modparam("rtpengine", "rtpengine_sock", "1 == udp:localhost:2223")
#modparam("rtpengine", "rtpengine_sock", "2 == udp:localhost:2224")
modparam("rtpengine", "setid_avp", "$avp(setid)")
modparam("rtpengine", "extra_id_pv", "$avp(extra_id)")

modparam("path", "use_received", 1)

# ----- ctl params -----
modparam("ctl", "binrpc", "unix:/var/run/kamailio_pcscf/kamailio_ctl")

# ----------------- Settings for Dispatcher ---------------
modparam("dispatcher", "list_file", "/etc/kamailio_pcscf/dispatcher.list")

# Dispatcher: Enable Failover-Support
modparam("dispatcher", "flags", 2)
# Dispatcher: Overwrite Destination address, if required.
modparam("dispatcher", "force_dst", 1)
# AVP's required for Fail-Over-Support:
#modparam("dispatcher", "dst_avp", "$avp(DISPATCHER_DST_AVP)")
#modparam("dispatcher", "grp_avp", "$avp(DISPATCHER_GRP_AVP)")
#modparam("dispatcher", "cnt_avp", "$avp(DISPATCHER_CNT_AVP)")
#modparam("dispatcher", "sock_avp", "$avp(DISPATCHER_SOCK_AVP)")

#modparam("dispatcher", "xavp_dst", "$avp(DISPATCHER_DST_AVP)")
#modparam("dispatcher", "xavp_dst_mode", 0)
#modparam("dispatcher", "xavp_ctx", "$avp(DISPATCHER_CNT_AVP)")
#modparam("dispatcher", "xavp_ctx_mode", 0)

# Try to recover disabled destinations every 15 seconds.
modparam("dispatcher", "ds_ping_interval", 15)
# Actively query the gateways:
modparam("dispatcher", "ds_probing_mode", 1)

# -- usrloc params --
#!ifdef DB_URL
#!ifdef DB_URL2
modparam("ims_usrloc_pcscf", "db_url", "cluster://cluster1")
#!else
modparam("ims_usrloc_pcscf", "db_url", DB_URL)
#!endif
modparam("ims_usrloc_pcscf", "db_mode", 0)
#!endif
#modparam("ims_usrloc_pcscf", "hashing_type", 2)
modparam("ims_usrloc_pcscf", "enable_debug_file", 0)
modparam("ims_usrloc_pcscf", "match_contact_host_port", 1)
modparam("ims_registrar_pcscf", "is_registered_fallback2ip", 1)
modparam("ims_registrar_pcscf", "ignore_reg_state", 1)
modparam("ims_registrar_pcscf", "ignore_contact_rxport_check", 1)
modparam("ims_registrar_pcscf", "pending_reg_expires", 30)
modparam("ims_registrar_pcscf", "subscription_expires", 36000)
modparam("ims_usrloc_pcscf", "expires_grace", 120)

#!ifdef WITH_REGINFO
modparam("ims_registrar_pcscf", "subscribe_to_reginfo", 1)
modparam("ims_registrar_pcscf", "publish_reginfo", 1)
modparam("ims_registrar_pcscf", "pcscf_uri", "sip:HOSTNAME")
#!else
modparam("ims_registrar_pcscf", "subscribe_to_reginfo", 0)
modparam("ims_registrar_pcscf", "publish_reginfo", 0)
#!endif

#!ifdef WITH_IPSEC
modparam("ims_ipsec_pcscf", "ipsec_listen_addr", IPSEC_LISTEN_ADDR)
modparam("ims_ipsec_pcscf", "ipsec_client_port", IPSEC_CLIENT_PORT)
modparam("ims_ipsec_pcscf", "ipsec_server_port", IPSEC_SERVER_PORT)
modparam("ims_ipsec_pcscf", "ipsec_spi_id_start", 4096)
modparam("ims_ipsec_pcscf", "ipsec_max_connections", IPSEC_MAX_CONN)
modparam("htable", "htable", "ipsec_clients=>size=8;autoexpire=600000;")
#!endif

#!ifdef WITH_RX
# -- CDP params --
modparam("cdp","config_file","/etc/kamailio_pcscf/pcscf.xml")
# -- diameter_rx params --
modparam("ims_qos", "rx_dest_realm", "PCRF_REALM")
#modparam("ims_qos", "rx_forced_peer", "pcrf.epc.mnc001.mcc001.3gppnetwork.org")
#modparam("ims_qos", "rx_forced_peer", "10.4.128.21")
#modparam("ims_qos", "rx_dest_realm", "NETWORKNAME")
modparam("ims_qos", "early_qosrelease_reason", "Sorry - QoS failed")
modparam("ims_qos", "confirmed_qosrelease_headers", "X-Reason: QoS failed\r\n")
modparam("ims_qos", "authorize_video_flow", 1)
modparam("ims_qos", "af_signaling_ip", RX_AF_SIGNALING_IP)
modparam("ims_qos", "include_rtcp_fd", 1)
modparam("ims_qos", "rx_auth_expiry", 36000)
#!endif

# -- pua params --
#!ifdef WITH_REGINFO
#!ifdef DB_URL
#!ifdef DB_URL2
modparam("pua", "db_url", "cluster://cluster1")
#!else
modparam("pua", "db_url", DB_URL)
#!endif
#!endif
#!endif

# -- ims_dialog params --
modparam("ims_dialog", "dlg_flag", FLT_DIALOG)
modparam("ims_dialog", "timeout_avp", "$avp(DLG_TIMEOUT_AVP)")
modparam("ims_dialog", "detect_spirals", 0)
modparam("ims_dialog", "profiles_no_value", "orig ; term")
#!ifdef DB_URL
#!ifdef DB_URL2
modparam("ims_dialog", "db_url", "cluster://cluster1")
#!else
modparam("ims_dialog", "db_url", DB_URL)
#!endif
modparam("ims_dialog", "db_mode", 0)
#!endif

#!ifdef CAPTURE_NODE
# Destination, where to send the traffic
modparam("siptrace", "duplicate_uri", CAPTURE_NODE)
# Trace all traffic
modparam("siptrace", "trace_on", 1)
modparam("siptrace", "trace_to_database", 0)
modparam("siptrace", "trace_flag", FLT_CAPTURE)
modparam("siptrace", "hep_mode_on", 1)
#!endif

# -- statistics params --
modparam("statistics", "variable", "register_success")
modparam("statistics", "variable", "register_failed")
modparam("statistics", "variable", "register_time")


####### Routing Logic ########

# Main SIP request routing logic
# - processing of any incoming SIP request starts with this route
route {
##!ifdef WITH_DEBUG
        xnotice("PCSCF: $rm $ru ($fu ($si:$sp) to $tu, $ci)\n");
##!endif

#!ifdef WITH_WEBSOCKET
        if (($Rp == MY_WS_PORT || $Rp == MY_WSS_PORT) && !(proto == WS || proto == WSS)) {
                xlog("L_WARN", "Websocket-request received on SIP/$Rp\n");
                sl_send_reply("403", "Forbidden - Websocket-request received on SIP/$Rp");
                exit;
        }
#!endif
        # per request initial checks
        route(REQINIT);

        # CANCEL processing
        if (is_method("CANCEL")) {
                if (t_check_trans()) {
                        t_relay();
                }
                exit;
        }

        if (is_method("NOTIFY") && (uri==myself)) {
                route(NOTIFY);
                exit;
        }

        # handle retransmissions
        if(t_precheck_trans()) {
                t_check_trans();
                exit;
        }
        t_check_trans();

        # Check for Re-Transmissions
        t_check_trans();

        # handle requests within SIP dialogs
        route(WITHINDLG);

        ### only initial requests (no To tag)

        if (is_method("UPDATE")) {
                send_reply("403","Forbidden - Target refresh outside dialog not allowed");
                break;
        }
        if (is_method("BYE|PRACK")) {
                send_reply("403","Forbidden - Originating subsequent requests outside dialog not allowed");
                break;
        }

        #Set DLG flag to track dialogs using dialog2
        if (!is_method("REGISTER|SUBSCRIBE"))
                setflag(FLT_DIALOG);

        loose_route();
#!ifdef WITH_SBC
        if (ds_is_from_list(DISPATCHER_LIST_SBC)) {
                if (is_method("INVITE")) {
                        if (is_present_hf("C-Params")) {
                                remove_hf("Contact");
                                remove_hf("C-Params");
                                append_hf("Contact: $ct;$hdr(C-Params)\r\n");
                        }
                        if ($route_uri =~ "sip:mo@.*") {
                                # prepend mo as user for record route
                                $avp(RR_CUSTOM_USER_AVP)="mo";
                                if (is_present_hf("P-Route")) {
                                        $du = $(hdr(P-Route){nameaddr.uri});
                                        remove_hf("P-Route");
                                        append_hf("Route: $hdr(P-Route)\r\n");
                                }
                                t_on_reply("SBC_GET_CPARAMS");
                        }
                        if ($route_uri =~ "sip:mt@.*") {
                                $du = $ru;
                                handle_ruri_alias();
                                if ($rc == 1) {
                                        setflag(FLT_NAT);
                                }
                                xlog("$$dP => $(dP{s.tolower}) ($du)\n");
                                # prepend mo as user for record route
                                $avp(RR_CUSTOM_USER_AVP)="mt";
                                if ($(dP{s.tolower}) == "tls") {
                                        route(ENC_SRTP);
                                } else if ($(dP{s.tolower}) == "ws") {
                                        # Even WSS is incorrectly shown as WS
                                        route(ENC_WSS_RTP);
                                } else if ($(dP{s.tolower}) == "wss") {
                                        route(ENC_WSS_RTP);
                                } else {
                                        route(ENC_RTP);
                                }
                                # Handle NAT
                                route(NATMANAGE);
                                # Handle Mobile Terminated requests
                                route(MT);
                        }
                }
                if (is_method("REGISTER")) {
                        append_hf("Path: <sip:term@$Ri:$Rp;lr>\r\n");
                }
        } else {
#!endif
                if ($route_uri =~ "sip:term@.*") {
#!ifdef WITH_SBC
#!ifdef WITH_SBC_CALL
                        if (is_method("INVITE")) {
#!endif
                                if (!strempty($(ct{tobody.params}))) {
                                        append_hf("C-Params: $(ct{tobody.params})\r\n");
                                }

                                append_hf("SBC: mt\r\n");
                                # Do some Round-Robin on the SBC's
                                t_on_failure("SBC_failure");
                                # Choose an SBC to send the call to:
                                if (!ds_select_dst(DISPATCHER_LIST_SBC, "4")) {
                                        send_reply("503", "Service Unavailable (SBC failure)");
                                        exit;
                                }
#!ifdef WITH_SBC_CALL
                        }
#!endif
#!else
                        handle_ruri_alias();
                        if ($dP == "tls") {
                                route(ENC_SRTP);
                        } else if ($dP == "ws") {
                                route(ENC_WS_RTP);
                        } else if ($dP == "wss") {
                                route(ENC_WSS_RTP);
                        } else {
                                route(ENC_RTP);
                        }
                        if ($rc == 1) {
                                setflag(FLT_NAT);
                        }
                        # Handle NAT
                        route(NATMANAGE);
                        # prepend mo as user for record route
                        $avp(RR_CUSTOM_USER_AVP)="mt";
                        route(MT);
#!endif
                } else {
                        force_rport();
                        if(is_method("INVITE|SUBSCRIBE|UPDATE|REGISTER")) {
                                add_contact_alias();
                        }
                        setflag(FLT_NAT);

                        if (is_method("REGISTER")) {
                                route(REGISTER);
                                exit;
                        }

                        # prepend mo as user for record route
                        $avp(RR_CUSTOM_USER_AVP)="mo";
                        # Set Flag for MO:
                        setflag(FLT_MOBILE_ORIG);
                        # Increase timer for inbound requests, we may have to do failover:
                        t_set_fr(120000, 30000);
                        route(MO);
                        if (is_method("INVITE")) {
                                # SRTP in RTP übersetzen
                                if ($pr == "tls") {
                                        route(DEC_SRTP);
                                } else if ($pr == "ws") {
                                        route(DEC_WS_RTP);
                                } else if ($pr == "wss") {
                                        route(DEC_WSS_RTP);
                                } else {
                                        route(DEC_RTP);
                                }
                                # Handle NAT
                                route(NATMANAGE);
                        }
#!ifdef WITH_SBC
#!ifdef WITH_SBC_CALL
                        if (is_method("INVITE")) {
#!endif
                                # Apply changes to this message
                                msg_apply_changes();
                                # Copy Route-Header:
                                append_hf("P-Route: $hdr(Route)\r\n");
                                if (!strempty($(ct{tobody.params}))) {
                                        append_hf("C-Params: $(ct{tobody.params})\r\n");
                                }
                                append_hf("SBC: mo\r\n");
                                # Do some Round-Robin on the SBC's
                                t_on_failure("SBC_failure");
                                # Choose an SBC to send the call to:
                                if (!ds_select_dst(DISPATCHER_LIST_SBC, "4")) {
                                        send_reply("503", "Service Unavailable (SBC failure)");
                                        exit;
                                }
#!endif
#!ifdef WITH_SBC_CALL
                        }
#!endif
                }
#!ifdef WITH_SBC
        }
#!endif
        if (is_method("INVITE|SUBSCRIBE")) {
                # record routing for dialog forming requests (in case they are routed)
                record_route();
        }

        # Forward request:
        route(RELAY);
        exit;
}

route[SBC_GET_CPARAMS] {
        if (!strempty($(ct{tobody.params}))) {
                append_hf("C-Params: $(ct{tobody.params})\r\n");
        }
}

# Per SIP request initial checks
route[REQINIT] {
        # Reply to OPTIONS:
        if (is_method("OPTIONS") && (uri==myself)) {
                options_reply();
                exit;
        }

        $var(used) = 1 - ($stat(free_size) / $stat(total_size));
        xlog("L_DBG", "Mem: Total $stat(total_size), Free $stat(free_size) [$var(used)% used]\n");
        if ($var(used) > 95) {
                send_reply("503", "Server overloaded");
                exit;
        }

        # Trace this message
#!ifdef CAPTURE_NODE
        sip_trace();
        setflag(FLT_CAPTURE);
#!endif

#!ifdef WITH_ANTIFLOOD
        # flood dection from same IP and traffic ban for a while
        # be sure you exclude checking trusted peers, such as pstn gateways
        # - local host excluded (e.g., loop to self)
        if (!has_totag() && (src_ip!=myself) && !ds_is_from_list())
        {
                if($sht(ipban=>$si)!=$null)
                {
                        # ip is already blocked
                        xlog("request from blocked IP - $rm from $fu (IP:$si:$sp)\n");
                        xlog("Blocking traffic from $si\n");
                        exit;
                }
                if (!pike_check_req()) {
                        xlog("L_ALERT","ALERT: pike blocking $rm from $fu (IP:$si:$sp, $ua)\n");
                        xlog("Blocking traffic from $si\n");
                        $sht(ipban=>$si) = 1;
                        exit;
                }
        }
        if ((uri == myself) && is_method("INVITE|REGISTER")) {
                xlog("L_ALERT","ALERT: Request to myself: $ru from $fu (IP:$si:$sp, $ua), Blocking\n");
                xlog("Blocking traffic from $si\n");
                $sht(ipban=>$si) = 1;
                exit;
        }
#!endif

        if (!mf_process_maxfwd_header("10")) {
                sl_send_reply("483","Too Many Hops");
                exit;
        }

        if(!sanity_check("1511", "7")) {
                xlog("Malformed SIP message from $si:$sp\n");
                exit;
        }

        # Check for shutdown mode:
        if (!has_totag() && ($sel(cfg_get.system.shutdownmode) > 0)) {
                send_reply("503", "Server shutting down");
                exit;
        }
#!ifdef WITH_IPSEC
        if (!is_method("REGISTER")) {
        ipsec_forward("location");
  }
#!endif
        # Ignore Re-Transmits:
        if (t_lookup_request()) {
                exit;
        }

        if (is_method("INVITE|REGISTER")) {
                send_reply("100", "Trying");
        }
}

route[RELAY] {
        if (!t_relay()) {
                sl_reply_error();
        }
        exit;
}

# Handle requests within SIP dialogs
route[WITHINDLG] {
        if (has_totag()) {
                xnotice("Within DLG\n");
                # sequential request withing a dialog should
                # take the path determined by record-routing
                if (loose_route()) {
                        xnotice("Within loose route\n");
                        if(!isdsturiset()) {
                                handle_ruri_alias();
                                #if ($rc == 1) {
                                #       $ru = "sip:" + $rU + "@" + $dd + ":" + $dp + ";transport=" + $rP;
                                #}
                        }

                        if ( is_method("ACK") && ($sht(contact=>$ci) != $null) ) {
                                xlog("Contact of Reply: $sht(contact=>$ci) ($ci)\n");
                                $ru = $sht(contact=>$ci);
                        }

                        if ($route_uri =~ "sip:mt@.*") {
                                route(MT_indialog);
                        } else {
                                route(MO_indialog);
                        }

                        # Handle NAT
                        route(NATMANAGE);

                        route(RELAY);
                        exit;
                }
                if ( is_method("ACK") ) {
                        xlog("Contact of Reply: $T_rpl($ct)\n");
                        if ( t_check_trans() ) {
                                # no loose-route, but stateful ACK;
                                # must be an ACK after a 487
                                # or e.g. 404 from upstream server
                                t_relay();
                                exit;
                        } else {
                                # ACK without matching transaction ... ignore and discard
                                exit;
                        }
                }
                sl_send_reply("404","Not here");
                exit;
        }
}

######################################################################
# Negative replies to REGISTER requests:
######################################################################
failure_route[SBC_failure] {
#!ifdef WITH_IPBLOCK
        if (is_method("REGISTER") && !ds_is_from_list()) {
                if (t_check_status("403|[5-6][0-9][0-9]")) {
                        if ($sht(failedauth=>$si) != $null)
                                $sht(failedauth=>$si) = $sht(failedauth=>$si) + 1;
                        else
                                $sht(failedauth=>$si) = 1;
                        if ($sht(failedauth=>$si) > 10) {
                                xlog("L_ALERT","ALERT: blocking $rm from $fu (IP:$si:$sp), more than 5 failed auth requests!\n");
                                xlog("Blocking traffic from $si\n");
                                $sht(ipban=>$si) = 1;
                        }
                }
        }
#!endif
        # Choose another gateway, in case we
        # - get a local generated "408"
        # - receive a 5xx or 6xx reply from the proxy.
        if (t_branch_timeout() || t_check_status("[5-6]..")) {
                if (ds_next_dst()) {
                        # Do Failover in case problems:
                        t_on_failure("SBC_failure");
                        t_relay();
                } else {
                        # Add a header, to indicate the phone should try again in 30 seconds.
                        append_hf("Retry-After: 30\r\n");
                        send_reply("504", "Server Time-Out");
                }
                exit;
        }
}

# Notify Route:                                                     #
#####################################################################
route[NOTIFY]
{
        xlog("L_DBG", "IMS: INSIDE NOTIFY\n");
        if (reginfo_handle_notify("location")) {
                send_reply("200","OK - P-CSCF processed notification");
                break;
        } else {
                t_reply("500","Error encountered while processing notification");
                break;
        }
}

#!ifdef WITH_NATPING
# NATPING Route:                                                    #
#####################################################################
route[NATPING] {
        route(preload_pcscf);

        sht_lock("natping=>natpinglock");
        sht_iterator_start("nat_iterator", "natping");
        while(sht_iterator_next("nat_iterator")) {
                xlog("OPTIONS to $shtitval(nat_iterator) via $shtitkey(nat_iterator)...\n");
                $uac_req(method) = "OPTIONS";
                $uac_req(ruri) = $shtitval(nat_iterator);
                $uac_req(furi) = PCSCF_URL;
                $uac_req(sock) = $sht(natpingfrom=>$shtitkey(nat_iterator));
                $uac_req(turi) = $shtitval(nat_iterator);
                $uac_req(ouri) = $shtitkey(nat_iterator);
                $uac_req(evroute) = 1;
                uac_req_send();
        }
        sht_iterator_end("nat_iterator");
        sht_unlock("natping=>natpinglock");
}

event_route[uac:reply] {
##!ifdef WITH_DEBUG
        xlog("request sent to $uac_req(ruri) completed with code: $uac_req(evcode), Type $uac_req(evtype)\n");
##!endif
        if (($uac_req(evtype) != 1) || ($uac_req(evcode) != 200)) {
                if ($sht(natpingfail=>$uac_req(ouri)) == $null) {
                        $sht(natpingfail=>$uac_req(ouri)) = 1;
                } else {
                        $sht(natpingfail=>$uac_req(ouri)) = $sht(natpingfail=>$uac_req(ouri)) + 1;
                }
                xlog("  request sent to $uac_req(ruri): Fail Counter is $sht(natpingfail=>$uac_req(ouri))\n");
                if ($sht(natpingfail=>$uac_req(ouri)) > 30) {
                        if ($(uac_req(ouri){uri.transport}) == "tcp") {
                                $var(alias) = "alias="+$(uac_req(ouri){uri.host})+"~"+$(uac_req(ouri){uri.port})+"~2";
                        } else if ($(uac_req(ouri){uri.transport}) == "tls") {
                                $var(alias) = "alias="+$(uac_req(ouri){uri.host})+"~"+$(uac_req(ouri){uri.port})+"~3";
                        } else {
                                $var(alias) = "alias="+$(uac_req(ouri){uri.host})+"~"+$(uac_req(ouri){uri.port})+"~1";
                        }
                        xlog("  Unregistering $uac_req(ruri);$var(alias)\n");
                        setdebug("9");
#!ifdef WITH_IPSEC
                        #if ($sht(ipsec_clients=>$(T_req($ct){nameaddr.uri})) != $null) {
                                #ipsec_destroy("location");
                                ipsec_destroy_by_contact("location", "$uac_req(ruri);$var(alias)", "$(uac_req(ouri){uri.host})", "$(uac_req(ouri){uri.port})");
                        #}
#!endif
                        pcscf_unregister("location", "$uac_req(ruri);$var(alias)", "$(uac_req(ouri){uri.host})", "$(uac_req(ouri){uri.port})");
                        resetdebug();
                        sht_lock("natping=>natpinglock");
                        $sht(natping=>$uac_req(ouri)) = $null;
                        sht_unlock("natping=>natpinglock");
                        $sht(natpingfail=>$uac_req(ouri)) = $null;

                        sht_lock("natpingfrom=>natpingfromlock");
                        $sht(natpingfrom=>$uac_req(ouri)) = $null;
                        sht_unlock("natpingfrom=>natpingfromlock");
                        $sht(natpingfail=>$uac_req(ouri)) = $null;
                }
        } else {
                $sht(natpingfail=>$uac_req(ouri)) = $null;
        }
}

event_route[htable:expired:natping] {
    xlog("natping record expired $shtrecord(key) => $shtrecord(value)\n");
}

event_route[htable:mod-init] {
        xlog("event_route[htable:mod-init] \n");
}

route[preload_pcscf] {
        if ($shv(preload_pcscf) == 1) return;
        $shv(preload_pcscf) = 1;

        sql_query("pcscf", "select aor, received, received_port, received_proto from location;", "resultset");
        xlog("Preloading NAT-PING. Rows: $dbr(resultset=>rows)\n");
        if($dbr(resultset=>rows)>0) {
                $var(i) = 0;
                while($var(i) < $dbr(resultset=>rows)) {
                        $var(ouri) = "sip:"+$dbr(resultset=>[$var(i),1])+":"+$dbr(resultset=>[$var(i),2]);
                        if ($dbr(resultset=>[$var(i),3]) == 2) {
                                $var(ouri) = $var(ouri)+";transport=tcp";
                        } else if ($dbr(resultset=>[$var(i),3]) == 3) {
                                $var(ouri) = $var(ouri)+";transport=tls";
                        }
                        $var(noalias) = $(dbr(resultset=>[$var(i),0]){re.subst,/^(.*);alias=.*/\1/}{nameaddr.uri});
                        # xlog("$$var(noalias) => $var(noalias) (via $var(ouri))\n");
                        sht_lock("natping=>natpinglock");
                        $sht(natping=>$var(ouri)) = $var(noalias);
                        sht_unlock("natping=>natpinglock");
                        $var(i) = $var(i) + 1;
                }
        }
        sql_result_free("resultset");
}
#!endif

#!ifdef WITH_XMLRPC
include_file "route/xmlrpc.cfg"
#!endif
#!ifdef WITH_WEBSOCKET
include_file "route/websocket.cfg"
#!endif
include_file "route/register.cfg"
include_file "route/rtp.cfg"
include_file "route/mo.cfg"
include_file "route/mt.cfg"
```


```
root@vo-i:/etc/kamailio_icscf# uname -a
Linux vo-i 4.15.0-124-generic #127-Ubuntu SMP Fri Nov 6 10:54:43 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```


```
root@vo-i:~# lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.5 LTS
Release:        18.04
Codename:       bionic
```



I'm Also using Open5gs V1.3.0 and 4g branch of your configs (https://github.com/herlesupreeth/Open5gs_Config)
All CORE + IMS components are on OpenStack with 4v CPU, 16GB RAM, 80GB disk and RAN part (SRSLTE) on another machine with Native Ubuntu, 8GB RAM, Core i7, 128 SSD , SDR is LIMESDR , USB3

**Comment by Mopoma95** [comment]
These Are also UE images if it helps!
Its the same as when i connect to commercial Volte network

![Screenshot_20201202-180153_Call settings](https://user-images.githubusercontent.com/38971099/100993932-41fffb80-356b-11eb-8d7f-389fc2d04da1.jpg)
![Screenshot_20201202-180200_Settings](https://user-images.githubusercontent.com/38971099/100993935-43312880-356b-11eb-885b-f73c9d757d73.jpg)

**Comment by herlesupreeth** [comment]
I dont see any misconfiguration in your setup, everything looks good. Regarding #5 (comment) I had a look on a general basis and everything looks good to me (also from the traces you shared earlier UE was authenticated in EPC so I dont find anything wrong with SIM but I cannot say for sure).

Now try once again running eNB + EPC + IMS and try UE attachment and send me a wireshark trace

**Comment by Mopoma95** [comment]
Thanks for your examination,
For sim i was talking about VOLTE part, i know the HSS has accepted the UE and has internet, But for Volte i think the sim programming VOLTE tab may have wrong configuration and for this wrong VOLTE part of Simcard configuration, we do not have VOLTE on UE or the PCSCF reject the acceptance of SIP!( Maybe, and if u looked at #5 VOLTE tab and its right. the problem is somewhere else!)



For Wireshark I traced it and it has 102MB then i compress it and now 25MB but Github couldnt accept file over 10MB!
So i uploaded it at  https://s17.picofile.com/file/8416172842/VOLTE.rar.html 
by clicking at red boxes u can download the file or use this link please:
https://s17.picofile.com/d/8416172842/b617882d-3fe2-40c6-a95e-b29a22e94a9e/VOLTE.rar

**Comment by Mopoma95** [comment]
Dear @herlesupreeth 
Have u tried to check the wireshark?

Thanks 
Best regards

**Comment by herlesupreeth** [comment]
If you dont mind, can you upload to google drive or mega and send me the link? I kind that website bit sketchy

**Comment by Mopoma95** [comment]
Sure, https://drive.google.com/file/d/1UJaEzFBMgxIwdO-D0vfZpEXunr58ABdS/view?usp=sharing
Please confirm if u can access that

**Comment by herlesupreeth** [comment]
I can. Thank you

**Comment by herlesupreeth** [comment]
I just dont understand, why the heck is your pcap 11GB.. this is definitely strange

**Comment by herlesupreeth** [comment]
I definitely suggest you to use this repo - https://github.com/herlesupreeth/docker_open5gs/tree/4g, create a new openstack VM, run as per the README in this specific link (not the master branch)

**Comment by Mopoma95** [comment]
> I just dont understand, why the heck is your pcap 11GB.. this is definitely strange

Its not 11GB!
Its just 110 MB

**Comment by Mopoma95** [comment]
The time was long, I connect and disconnect the UE triple times! that's why it get large

I will try and did try the Docker before, But Now i'm gonna work with The native one


Do u think it might related to open5GS? I mean the Diameter of Open5gs or sth else which @acetcom can help?


Thanks for ur understanding
Best regards

**Comment by herlesupreeth** [comment]
Its definitely not the fault in open5gs or kamailio IMS, looks like a networking issue

the docker method I mentioned before will get you a complete EPC + IMS setup running in less than 30-40 min and that too using 3-4 commands

**Comment by Mopoma95** [comment]
The problem with Docker is with the links, For example JDK and some other things which is sensorded here! and is not accessible!
But with VOLTE instruction i can Download them manually and can turn the sensors off!


Any comment on the wireshark?
I will try again with the VOLTE instruction and tell u if something new happened!

Thanks 
Best regards

**Comment by herlesupreeth** [comment]
As i see in the volte.pcap, FHoSS is sending Authorization Reject

![image](https://user-images.githubusercontent.com/4115839/101147468-b47ce400-361c-11eb-8cc4-ed48615c9ce3.png)

Looking at this screenshot https://user-images.githubusercontent.com/38971099/100891729-95008280-3487-11eb-82cf-34b0cbdb1c5c.png, I can tell you skipped one step in the below section of VoLTE tutorial i.e. And, finally add these IMPUs as implicit set of IMSI derived IMPU in HSS blah blah, please do this part

![image](https://user-images.githubusercontent.com/4115839/101147796-1d645c00-361d-11eb-8bbc-e3e12697990b.png)

**Comment by Mopoma95** [comment]
Dear @herlesupreeth 
Thanks you so much for giving me your time,

I did those steps, you can see in pics if i'm not mistaken.
one thing that might be the problem is the SQN!!!
I did not set that in HSS but for the sim i created for your tutorial's test (00101 ... ) it automatically set that to sth like ...161 and i can't change it using subscriber settings!
but i followed ur tutorial and i set that to 0000000210 !

So now i have a question about that, Is that might be the problem or as u said i skipped those steps u mentioned?
If the Problem is about SQN, can i delete the HSS database info and make it again? doesn't it make trouble in other components?

As I checked, I do not have access to set SQN even to new subscriber!!! I'm using open5gs-2.0.0 for webUI.
How do i know the complete SQN which has been set in Subscriber? I mean know the ... in ...161 which has been subscribed in database?


![Screenshot (30)](https://user-images.githubusercontent.com/38971099/101191609-241bbf00-366f-11eb-9529-b1a90d074a2a.png)
![Screenshot (31)](https://user-images.githubusercontent.com/38971099/101191615-25e58280-366f-11eb-9fb9-77618f19171c.png)
![Screenshot (32)](https://user-images.githubusercontent.com/38971099/101191620-2716af80-366f-11eb-92e6-376e30143b64.png)
![Screenshot (33)](https://user-images.githubusercontent.com/38971099/101191624-27af4600-366f-11eb-9ca8-b0a484e514e3.png)
![Screenshot (34)](https://user-images.githubusercontent.com/38971099/101191627-28e07300-366f-11eb-9822-a4211711780f.png)
![Screenshot (35)](https://user-images.githubusercontent.com/38971099/101191629-29790980-366f-11eb-953f-90f8a1bf4809.png)




Thanks
Best regards

**Comment by Mopoma95** [comment]
Dear @herlesupreeth 
Hi, and Thanks for your guidance

I have successfully connected the UE to VOLTE , Thank u so much 

but it seams to be some problems for call!

Can u check the Logs?


[Volte.zip](https://github.com/herlesupreeth/Open5gs_Config/files/5650508/Volte.zip)

Thanks
Best regards

**Comment by Mopoma95** [comment]
Thanks again For your Help,

I'm gonna close this, If i face the New problem, will open new issue


Thanks again
Best regards
